core/common/tarutil.cpp

```diff
@@ -38,6 +39,96 @@ namespace fc::common {
     }
   }
 
+  outcome::result<void> zipTar(const std::string &input_path,
+                               const std::string &output_path) {
+    if (!fs::exists(input_path)) {
+      logger->error("Zip tar: {} doesn't exists", input_path);
+      return TarErrors::kCannotZipTarArchive;
+    }
+    if (!fs::is_directory(input_path)) {
+      logger->error("Zip tar: {} is not a directory", input_path);
+      return TarErrors::kCannotZipTarArchive;
+    }
+    if (fs::exists(output_path) && !fs::is_regular_file(output_path)) {
+      logger->error("Zip tar: {} is not a file", output_path);
+      return TarErrors::kCannotZipTarArchive;
+    }
+
+    auto a = ffi::wrap(archive_write_new(), archive_write_free);
+    archive_write_set_format_v7tar(a.get());
+    int r = archive_write_open_filename(a.get(), output_path.c_str());
+    if (r < ARCHIVE_OK) {
+      if (r < ARCHIVE_WARN) {
+        logger->error("Zip tar: {}", archive_error_string(a.get()));
+        return TarErrors::kCannotZipTarArchive;
+      } else {
+        logger->warn("Zip tar: {}", archive_error_string(a.get()));
+      }
+    }
+    std::function<outcome::result<void>(const fs::path &, const fs::path &)>
+        zipDir = [&](const fs::path &absolute_path,
+                     const fs::path &relative_path) -> outcome::result<void> {
+      struct archive_entry *entry;
+      struct stat st {};
+      char buff[8192];
+      fs::directory_iterator dir_iter(absolute_path), end;
+
+      while (dir_iter != end) {
+        auto type = AE_IFREG;
+        if (fs::is_directory(dir_iter->path())) {
+          if (fs::directory_iterator(dir_iter->path())
+              != fs::directory_iterator()) {
+            OUTCOME_TRY(zipDir(dir_iter->path(),
+                               relative_path / dir_iter->path().filename()));
+            ++dir_iter;
+            continue;
+          }
+          type = AE_IFDIR;
+        }
+        if (stat(dir_iter->path().c_str(), &st) < 0) {
+          return TarErrors::kCannotZipTarArchive;
+        }
+        entry = archive_entry_new();
+        archive_entry_set_pathname(
+            entry, (relative_path / dir_iter->path().filename()).c_str());
+        archive_entry_set_size(entry, st.st_size);
+        archive_entry_set_filetype(entry, type);
+        archive_entry_set_perm(entry, 0644);
+        r = archive_write_header(a.get(), entry);
+        if (r < ARCHIVE_OK) {
+          if (r < ARCHIVE_WARN) {
+            logger->error("Zip tar: {}", archive_error_string(a.get()));
+            return TarErrors::kCannotZipTarArchive;
+          } else {
+            logger->warn("Zip tar: {}", archive_error_string(a.get()));
+          }
+        }
+        if (type == AE_IFREG) {
+          int fd = open(dir_iter->path().c_str(), O_RDONLY);
```

Comment:
Either use raii type (`PieceData`?) or `gsl::finally` to close `fd`

Reply:
Changed to fstream